lib/improver/nbhood/square_kernel.py

```diff
@@ -406,67 +454,79 @@ def mean_over_neighbourhood(self, cube, cells_x, cells_y, nan_masks):
+
             with np.errstate(invalid='ignore', divide='ignore'):
                 cube.data = (neighbourhood_total.astype(float) /
                              neighbourhood_area.astype(float))
+                cube.data[~np.isfinite(cube.data)] = np.nan
         elif self.sum_or_fraction == "sum":
             cube.data = neighbourhood_total.astype(float)
 
-        cube.data[nan_masks[0].astype(bool)] = np.NaN
         return cube
 
     @staticmethod
-    def _set_up_cubes_to_be_neighbourhooded(cube, mask_cube=None):
+    def set_up_cubes_to_be_neighbourhooded(cube, mask_cube=None):
         """
-        Set up a cubelist containing either the input cube, or the input cube
-        and a mask cube.
+        Set up a cube ready for neighourhooding the data.
 
         Args:
             cube (Iris.cube.Cube):
-                Cube that will be checked for whether the data is masked.
+                Cube that will be checked for whether the data is masked
+                or nan. The cube should contain only x and y dimensions,
+                so will generally be a slice of a cube.
 
         Keyword Args:
             mask_cube (Iris.cube.Cube):
-                Cube containing the array to be used as a mask.
+                Input Cube containing the array to be used as a mask.
 
         Returns:
-            cubes_to_sum (Iris.cube.CubeList):
-                CubeList containing either the input cube, or the input cube
-                and a mask cube.
+            cube (Iris.cube.Cube):
+                Cube with masked or NaN values set to 0.0
+            mask (Iris.cube.Cube):
+                Cube with masked or NaN values set to 0.0
+            nan_array (np.array):
+                numpy array to be used to set the values within
+                the data of the output cube to be NaN.
         """
+        # Set up mask_cube
+        if not mask_cube:
+            mask = cube.copy()
+            mask.data[::] = 1.0
+        else:
+            mask = mask_cube
         # If there is a mask, fill the data array of the mask_cube with a
         # logical array, logically inverted compared to the integer version of
         # the mask within the original data array.
-        if isinstance(cube.data, np.ma.MaskedArray) or mask_cube:
-            if not mask_cube:
-                mask_cube = cube.copy()
-                mask_cube.data = np.logical_not(cube.data.mask.astype(int))
-            mask_cube.rename('mask_data')
-            if np.ma.is_masked(cube.data):
-                cube.data = cube.data.data
-            cube = iris.util.squeeze(cube)
-            cube.data = (cube.data * mask_cube.data.squeeze()).astype(
-                cube.data.dtype)
-            cubes_to_sum = iris.cube.CubeList([cube, mask_cube])
-        else:
-            cubes_to_sum = iris.cube.CubeList([cube])
-        return cubes_to_sum
+        if isinstance(cube.data, np.ma.MaskedArray):
+            index = np.where(cube.data.mask.astype(int) == 1)
```

Comment:
Does this work if it is just `np.where(cube.data.mask)`?

Reply:
Possibly but since the values could be floats or Booleans I would rather explicitly type this as int for the comparison.

Reply:
Fair enough